tq/adapterbase.go

```diff
@@ -82,19 +90,58 @@ func (a *adapterBase) Begin(maxConcurrency int, cb TransferProgressCallback, com
 	return nil
 }
 
-func (a *adapterBase) Add(t *Transfer) {
-	tracerx.Printf("xfer: adapter %q Add() for %q", a.Name(), t.Object.Oid)
-	a.jobChan <- t
+type job struct {
+	T *Transfer
+
+	listeners []chan<- TransferResult
+	wg        *sync.WaitGroup
+}
+
+func (j *job) Done(err error) {
+	for _, l := range j.listeners {
+		l <- TransferResult{j.T, err}
+	}
+
+	j.wg.Done()
+}
+
+func (a *adapterBase) Add(transfers ...*Transfer) <-chan TransferResult {
+	results := make(chan TransferResult, len(transfers))
+
+	listeners := []chan<- TransferResult{results}
+	if a.outChan != nil {
+		listeners = append(listeners, a.outChan)
+	}
+
+	a.jobWait.Add(len(transfers))
+
+	go func() {
+		defer close(results)
+
+		for _, t := range transfers {
+			// BUG(taylor): End() is race-y here, and can close
+			// jobChan before we want it to
+			a.jobChan <- &job{t, listeners, a.jobWait}
```

Comment:
Doh! Sorry, I thought you were asking about my use of `a.jobWait.Wait()` in `Done()`, which I added after I added that `// BUG(taylor)` comment while debugging. Looks like I just forgot to remove that comment, which I did in ce08f661. I think we just found out in a round-about way that we're on the same page!